src/rdrand.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright 2018 Developers of the Rand project.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Implementation for SGX using RDRAND instruction
+use crate::Error;
+use core::arch::x86_64::_rdrand64_step;
+use core::mem;
+use core::num::NonZeroU32;
+
+#[cfg(not(target_feature = "rdrand"))]
+compile_error!("enable rdrand target feature!");
+
+// Recomendation from "Intel® Digital Random Number Generator (DRNG) Software
+// Implementation Guide" - Section 5.2.1 and "Intel® 64 and IA-32 Architectures
+// Software Developer’s Manual" - Volume 1 - Section 7.3.17.1.
+const RETRY_LIMIT: usize = 10;
+
+fn get_rand_u64(data: &mut u64) -> Result<(), Error> {
+    for _ in 0..RETRY_LIMIT {
+        unsafe {
+            if _rdrand64_step(data) == 1 {
+                return Ok(());
+            }
+        };
+    }
+    Err(Error::UNKNOWN)
+}
+
+fn fill_small_slice(slice: &mut [u8]) -> Result<(), Error> {
+    if slice.len() > 0 {
+        let mut data = unsafe { mem::uninitialized() };
+        get_rand_u64(&mut data)?;
+        slice.copy_from_slice(&data.to_ne_bytes()[..slice.len()])
+    }
+    Ok(())
+}
+
+pub fn getrandom_inner(dest: &mut [u8]) -> Result<(), Error> {
+    let (prefix, aligned, suffix) = unsafe { dest.align_to_mut() };
```

Comment:
Citing `align_to_mut` docs:

>The method does a best effort to make the middle slice the greatest length possible for a given type and input slice, but only your algorithm's performance should depend on that, not its correctness.

In other words strictly speaking it does not guarantee that `prefix` and `suffix` are smaller or equal to 8 bytes and you rely on it in `fill_small_slice`. Looking at generated assembly we can see that `slice_index_len_fail` is still present, so compiler can't prove that their length is smaller or equal to 8 bytes.

One option is to explicitly construct slices without relying on `std` method.

Reply:
I decided just to change `get_rand_unaligned` to support slices longer than 8 bytes, so we are now properly relying on `align_to_mut` for performance rather than correctness. Furthermore, the [generated assembly](https://rust.godbolt.org/z/wcFE62) now:
  - Contains no `slice_index_len_fail` or panic hooks, so the compiler can elide all of the slice checks.
  - Inlines all the functions (as expected).